### PR TITLE
Add FSIM and FSIMc

### LINKED ISSUE #104: Add FSIM and FSIMc measure
**Is your feature request related to a problem? Please describe.**
First described in [this paper](https://www4.comp.polyu.edu.hk/~cslzhang/IQA/TIP_IQA_FSIM.pdf), it showed good performance on different IQA datasets

**Describe the solution you'd like**
Reimplement authors MATLAB version on PyTorch

### PR DESCRIPTION
**Pull request by zakajd**
Closes #104 

## Proposed Changes
  - PyTorch implementation of FSIM and FSIMc
       - 99% accuracy with original MATLAB results
       - Batch computations support
       - Gradient optimisation support (with small hack, because PyTorch will add support for complex tensors grads only in version 1.6.0)

Not ready yet. Tests are still in progress

### CODE REVIEW

**Comment by zakajd** [comment]
@denproc  @snk4tr  ready for review

File: piq/fsim.py [hunk 1]
```diff
@@ -0,0 +1,453 @@
...
+    radius = torch.sqrt(grid_x ** 2 + grid_y ** 2)
+
+    return ifftshift(1. / (1.0 + (radius / cutoff) ** (2 * n)))
+
+
+class FSIMLoss(_Loss):
+    r"""Creates a criterion that measures the FSIM or FSIMc for input :math:`x` and target :math:`y`.
+
+    In order to be considered as a loss, value `1 - clip(FSIM, min=0, max=1)` is returned. If you need FSIM value,
+    use function `fsim` instead.
```
Comment (snk4tr) [inline piq/fsim.py:379]: Like this note 👍

File: piq/fsim.py [hunk 2]
```diff
@@ -0,0 +1,453 @@
...
+            prediction: Tensor of prediction of the network.
+            target: Reference tensor.
+        Returns:
+            Value of FSIM loss to be minimized. 0 <= FSIM <= 1.
+        """
+        # All checks are done inside fsim function
+        score = self.fsim(prediction, target)
+
+        # Make sure value to be in [0, 1] range and convert to loss
+        return 1 - torch.clamp(score, 0, 1)
```
Comment (snk4tr) [inline piq/fsim.py:453]: Just curious whether we need to do something about the case when `torch.clamp(score, 0, 1)` is equal to 1 🤔 
Are there any cases then it may cause some problems?

File: piq/functional.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
+from typing import Tuple
+
+import torch
+
+
+def ifftshift(x: torch.Tensor):
+    """ Similar to np.fft.ifftshift but applies to PyTorch Tensors
+    """
```
Comment (snk4tr) [inline piq/functional.py:8]: If you intend to make a single like docstring, make it actually a single line 😄 

```suggestion
    r""" Similar to np.fft.ifftshift but applies to PyTorch Tensors """
```

File: piq/functional.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    shift = [-(ax // 2) for ax in x.size()]
+    return torch.roll(x, shift, tuple(range(len(shift))))
+
+
+def get_meshgrid(size: Tuple[int, int]) -> torch.Tensor:
+    """
```
Comment (snk4tr) [inline piq/functional.py:14]: We use raw docstring everywhere 
```suggestion
    r"""
```

File: piq/functional.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+        # Odd
+        y = torch.range(-(size[1] - 1) / 2, (size[1] - 1) / 2) / (size[1] - 1)
+    else:
+        # Even
+        y = torch.range(- size[1] / 2, size[1] / 2 - 1) / size[1]
+    return torch.meshgrid(x, y)
+
+
+def similarity_map(map_x: torch.Tensor, map_y: torch.Tensor, constant: float) -> torch.Tensor:
+    """ Compute similarity_map between two tensors using Dice-like equation.
```
Comment (snk4tr) [inline piq/functional.py:35]:
```suggestion
    r""" Compute similarity_map between two tensors using Dice-like equation.
```

File: piq/functional.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    Args:
+        map_x: Tensor with map to be compared
+        map_y: Tensor with map to be compared
+        constant: Used for numerical stability
+    """
+    return (2.0 * map_x * map_y + constant) / (map_x ** 2 + map_y ** 2 + constant)
+
+
+def gradient_map(x: torch.Tensor, kernels: torch.Tensor) -> torch.Tensor:
+    """ Compute gradient map for a given tensor and stack of kernels.
```
Comment (snk4tr) [inline piq/functional.py:46]:
```suggestion
    r""" Compute gradient map for a given tensor and a stack of kernels.
```

File: tests/test_fsim.py [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+@pytest.mark.parametrize(
+    "x,y,expectation,value",
+    [
+        (torch.rand(4, 3, 128, 128, 2), torch.rand(4, 3, 128, 128, 2), pytest.raises(AssertionError), None),
+        (torch.randn(4, 3, 128, 128), torch.randn(4, 3, 128, 128), pytest.raises(AssertionError), None),
+        (torch.zeros(4, 3, 128, 128), torch.zeros(4, 3, 128, 128), raise_nothing(), 1.0),
+        (torch.ones(4, 3, 128, 128), torch.ones(4, 3, 128, 128), raise_nothing(), 1.0),
+        (torch.rand(4, 3, 28, 28), torch.rand(4, 3, 28, 28), raise_nothing(), None),
+    ],
+)
```
Comment (snk4tr) [inline tests/test_fsim.py:52]: Nice 👍

File: piq/fsim.py [hunk 8]
```diff
@@ -0,0 +1,453 @@
...
+    mean_o = sum_o / x_energy
+
+    # Now calculate An(cos(phase_deviation) - | sin(phase_deviation)) | by
+    # using dot and cross products between the weighted mean filter response
+    # vector and the individual filter response vectors at each scale.
+    # This quantity is phase congruency multiplied by An, which we call energy.
+
+    # Extract even and odd convolution results.
+    E = E0[..., 0]
+    O = E0[..., 1]
```
Comment (denproc) [inline piq/fsim.py:284]: I would suggest to avoid `O` variable as it can be misinterpreted as `0`.

File: piq/fsim.py [hunk 9]
```diff
@@ -0,0 +1,453 @@
...
+        
+    Returns:
+        f = 1 / (1 + w/cutoff) ^ 2n
+    
+    Note:
+        The frequency origin of the returned filter is at the corners.
+    
+    """
+
+    assert (cutoff >= 0) and (cutoff <= 0.5), "Cutoff frequency must be between 0 and 0.5"
```
Comment (denproc) [inline piq/fsim.py:364]:
```suggestion
    assert 0 <= cutoff <= 0.5, "Cutoff frequency must be between 0 and 0.5"
```

File: piq/fsim.py [hunk 10]
```diff
@@ -0,0 +1,453 @@
...
+    assert n > 1, "n must be an integer >= 1"
```
Comment (denproc) [inline piq/fsim.py:365]:
```suggestion
    assert n > 1 and int(n) == n, "n must be an integer >= 1"
```

File: piq/functional.py [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+
+
+def get_meshgrid(size: Tuple[int, int]) -> torch.Tensor:
+    """
+    Args:
+        size: Shape of meshgrid to create
+    """
+    if size[0] % 2:
+        # Odd
+        x = torch.range(-(size[0] - 1) / 2, (size[0] - 1) / 2) / (size[0] - 1)
```
Comment (denproc) [inline piq/functional.py:20]: `torch.range` is deprecated in favour of `torch.arange`.

File: piq/fsim.py [hunk 12]
```diff
@@ -0,0 +1,453 @@
...
+
+    B, _, H, W = x.shape
+
+    # Fourier transform
+    imagefft = torch.rfft(x, 2, onesided=False)
+
+    filters = _construct_filters(x, scales, orientations, min_length, mult, sigma_f, delta_theta, k)
+
+    # Note rescaling to match power record ifft2 of filter
+    filters_ifft = torch.ifft(torch.stack([filters, torch.zeros_like(filters)], dim=-1), 2)[..., 0] * math.sqrt(H * W)
```
[the remark this answers is not in the document]
Comment (zakajd) [reply piq/fsim.py:252]: No, due to unknown reason when complex part consists of zeros result of `ifft` and `irfft` is different.

File: piq/fsim.py [hunk 13]
```diff
@@ -0,0 +1,453 @@
...
+    spread = []
+    for o in range(orientations):
+        angl = o * math.pi / orientations
+
+        # For each point in the filter matrix calculate the angular distance from
+        # the specified filter orientation.  To overcome the angular wrap-around
+        # problem sine difference and cosine difference values are first computed
+        # and then the atan2 function is used to determine angular distance.
+        ds = sintheta * math.cos(angl) - costheta * math.sin(angl)  # Difference in sine.
+        dc = costheta * math.cos(angl) + sintheta * math.sin(angl)  # Difference in cosine.
```
[the remark this answers is not in the document]
Comment (zakajd) [reply piq/fsim.py:202]: I'll leave `math`, as using `torch` operations will require casting scalars to tensors first, which is inconvenient.

File: piq/fsim.py [hunk 14]
```diff
@@ -0,0 +1,453 @@
...
+    y = y / float(data_range)
+    
+    # Rescale to [0, 255] range, because all constant are calculated for this factor
+    x = x * 255
+    y = y * 255
+    
+    # Apply average pooling
+    kernel_size = max(1, round(min(x.shape[-2:]) / 256))
+    x = torch.nn.functional.avg_pool2d(x, kernel_size, stride=2)
+    y = torch.nn.functional.avg_pool2d(y, kernel_size, stride=2)
```
[the remark this answers is not in the document]
Comment (zakajd) [reply piq/fsim.py]: Done
